sound/soc/sof/ipc4-topology.c

```diff
@@ -1485,6 +1485,53 @@ static int sof_ipc4_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget
 	return ret;
 }
 
+static int sof_ipc4_get_queue_id(struct snd_sof_widget *swidget, int bind_type,
+				 const char *peer_name)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_sof_pin_binding *pin_binding;
+	struct ida *queue_ida;
+
+	int num_pins;
+	int i;
+
+	if (swidget->num_pin_bindings > 0) {
+		for (i = 0; i < swidget->num_pin_bindings; i++) {
+			pin_binding = &swidget->pin_binding[i];
+
+			if (pin_binding->bind_type == bind_type &&
+			    !strcmp(pin_binding->wname, peer_name)) {
+				return pin_binding->queue_id;
+			}
+		}
+		dev_err(scomp->dev, "failed to find queue id from pin binding array for %s\n",
+			swidget->widget->name);
+		return -EINVAL;
+	}
+
+	if (bind_type == SOF_ROUTE_BIND_SOURCE) {
+		queue_ida = &swidget->source_queue_ida;
+		num_pins = swidget->num_source_pins;
+	} else {
+		queue_ida = &swidget->sink_queue_ida;
+		num_pins = swidget->num_sink_pins;
+	}
+
+	return ida_alloc_max(queue_ida, num_pins, GFP_KERNEL);
```

Comment:
We need to check num_pins before calling `ida_alloc_max(queue_ida, num_pins, GFP_KERNEL);`. It could be 0 if we set it = 0 in topology and it will cause kernel oops.
Suggestion:
```
if (num_pins <= 0)
        num_pins = 1;
```
Or
```
if (num_pins < 1) {
        dev_err(scomp->dev, "Invalid num_pins %d\n", num_pins);
        return -EINVAL;
}
```

Reply:
As @ranj063  suggested, `swidget->num_source_pins` and `swidget->num_sink_pins` are initialized to 1 in `widget_ready()`, and topology will override this value, but there is no reason to define them as zero, so I think we are good here.

Reply:
> As @ranj063 suggested, `swidget->num_source_pins` and `swidget->num_sink_pins` are initialized to 1 in `widget_ready()`, and topology will override this value, but there is no reason to define them as zero, so I think we are good here.

No, we can't rely on the topology. We should prevent kernel from oops as much as we can do.

Reply:
yeah, a bad guy may override this value and make it 0 with topology change, let's prevent the damage of this malicious action also.